qkernel/src/interrupt/x86_64/mod.rs

```diff
@@ -773,3 +792,111 @@ pub extern "C" fn TripleFaultHandler(sf: &mut PtRegs) {
     info!("\nTripleFaultHandler: at {:#x}\n{:#?}", sf.rip, sf);
     loop {}
 }
+
+#[cfg(feature = "cc")]
+#[no_mangle]
+pub extern "C" fn VmmCommunicationHandler(sf: &mut PtRegs, errorCode: u64) {
+    use crate::Kernel_cc::LOG_AVAILABLE;
+    let log_available = LOG_AVAILABLE.load(core::sync::atomic::Ordering::Acquire);
+    let from_user = sf.cs & 0x3 != 0;
+    if from_user {
+        CPULocal::Myself().SetMode(VcpuMode::Kernel);
+        let currTask = Task::Current();
+        let mut rflags = sf.eflags;
+        rflags &= !USER_FLAGS_CLEAR;
+        rflags |= USER_FLAGS_SET;
+        sf.eflags = rflags;
+        sf.ss |= 3;
+        currTask.SaveFp();
+    }
+    /* 
+    if log_available && errorCode != SVMExitDef::SVM_EXIT_IOIO {
+        info!(
+            "Enter VC Handler errorCode:{}, PtRegs:{:#x?}",
+            errorCode, sf
+        );
+    }
+    */
+    let mut res = VCResult::EsDecodeFailed;
+    let code_ptr = sf.rip;
+    let code_slice;
+    unsafe {
+        code_slice = slice::from_raw_parts_mut(code_ptr as *mut u8, 15);
+    }
+
+    let mut ins_length = 0;
+    let decoder = yaxpeax_x86::amd64::InstDecoder::default();
+    match decoder.decode_slice(code_slice) {
+        Ok(i) => {
+            match errorCode {
+                SVMExitDef::SVM_EXIT_CPUID => res = vc_handle_cpuid(sf),
+                SVMExitDef::SVM_EXIT_INVD => res = VCResult::EsUnsupported,
+                SVMExitDef::SVM_EXIT_MONITOR => res = VCResult::EsOk,
+                SVMExitDef::SVM_EXIT_MWAIT => res = VCResult::EsOk,
+                //mmio not supported yet
+                SVMExitDef::SVM_EXIT_NPF => res = VCResult::EsUnsupported,
+                _ => {
+                    let mut vcpuid = 0;
+                    if log_available {
+                        vcpuid = crate::qlib::kernel::asm::GetVcpuId();
+                    }
+                    let ghcb_option: &mut Option<GhcbHandle<'_>> = &mut *GHCB[vcpuid].lock();
+                    let ghcb = ghcb_option.as_mut().unwrap();
+                    ghcb.invalidate();
```

Comment:
Why we need to do `ghcb.invalidate` twice? [I thought you already did it before vm exit](https://github.com/QuarkContainer/Quark/pull/1269/files#diff-399eb8f5992748de1a0edb97ae7b78e68890f79f416496f46f84b3c9ef7eafc2R422)